src/common/HaskellDo/State.hs

```diff
@@ -131,11 +131,8 @@ update (ToolbarAction Toolbar.SavePackage) appState = do
     return newState
 
 update (ToolbarAction action) appState = do
+    newToolbarState <- Toolbar.update action (toolbarState appState)
     let cs = compilationState appState
-    cs' <- atRemote $ Compilation.update Compilation.GetLastProject cs
-    let ts = toolbarState appState
-    let ts' = ts { Toolbar.projectPath = Compilation.projectPath cs' }
-    newToolbarState <- Toolbar.update action ts'
     let newCompilationState = cs
             { Compilation.projectPath = Toolbar.projectPath newToolbarState
```

Comment:
I think that instead of having a shared state it is better to rethink the components, for example Toolbar maybe should only have the Toolbar concept itself, not the buttons of each action.

For instance, the compile button should be in the `View` of `Compilation`, not the toolbar. And stuff like that, I think that the components deserve some refactoring to organize them and convert them into something more fine-grained.

Shared state is not an option in my opinion, it scales badly and as we start adding more feature it will get out of control 😄